cluster-controller/src/main/java/io/strimzi/controller/cluster/operator/resource/KafkaSetOperator.java

```diff
@@ -32,8 +32,8 @@ public KafkaSetOperator(Vertx vertx, KubernetesClient client) {
     protected Future<ReconcileResult<Boolean>> internalPatch(String namespace, String name, StatefulSet current, StatefulSet desired) {
         StatefulSetDiff diff = new StatefulSetDiff(current, desired);
         if (diff.changesVolumeClaimTemplates()) {
-            log.warn("Ignoring change to volumeClaim");
-            desired.getSpec().setVolumeClaimTemplates(current.getSpec().getVolumeClaimTemplates());
+            log.warn("Changing storage type or size is not possible.");
```

Comment:
* The log message says the change is not possible, fine, but it should also make clear that the change is automatically reverted, so anyone seeing the warning understands that they don't need to do anything to revert it themselves.
* We've got exactly this same block of code in the Zk operator, with the logging, the reverting and the recomputation of the diff: We should factor out a common method. 
* Thinking longer term: I suspect therre's actually still more commonality between the two `internalPatch()` methods (Kafka and Zk cases). Really all our statefulsets are statefulsets-with-attached-storage, and (in the future) some (abstracted) rolling update logic. So maybe we should just squeeze all this logic into the `StatefulSetOperator` → `StatefulSetWithStorageOperator`, or something.

Reply:
> (Which is in the model)

Since the resiliency changes we don't really use the model in this way any more:

* Before the resiliency changes we would construct two models and then compute their differences, and then apply changes according to those differences. 
* After the resiliency changes we compute the `desired` resources from the model which we create from the cluster CM, and then (more or less) patch the resources with that desired state. 

It's not clear to me where you propose the model would figure out the storage differences: It's only in the statefulset `internalPatch()` that we have access to the current and desired states, so do you propose we build a model at this point, or do you propose we build a model of the storage earlier in the reconciliation?

Reply:
Do you think the current way is maintainable in the long term? I'm afraid that with adding more storage types and more different volumes (ConfigMaps, Secrets etc.), this whole think would become very unstable and costly to maintain.

I know you do not use the model this way anymore. But the model is the place where you can fix this by simply generating the correct desired StatefulSets and you don't have to do any patching later. I'm not sure whether the model has to be the one figuring what is the right storage to use is - it can be just passed from outside.

Additional point which I realised (but that is not a new problem) is that we will not recover the old storage type after some "special" events (such as deleting the whole SS). But that is probably a bit artificial (e.g. when the SS is gone ... are we sure the PVC are still there?), so I'm not sure how much effort we should put in that.

Reply:
I wonder if we should store the storage configs as annotations on the cluster CM, rather that trusting the state of the SS. That would mean it was more resilient to the SS being changed or deleted directly by the user, and it would be available to `fromConfigMap()` to compute the desired state right at the start of reconciliation. This is also a solution for #348.

Reply:
@tombentley Yep, that is pretty much the direction I was thinking about as well. I wanted to check first what are the different ways to update the configmap which would lead to deleting the annotations.